WIP: Impacts endpoint to retrieve stand treatment results

Foxbyte0101 opened the pull request:
@george-silva @roquebetioljr 

Here's the WIP of the endpoint. The endpoint logic is in impacts/views.py, and the database logic is in a function at the bottom of impacts/services.py. 

For the services.py function, I need more specific details about what specific columns/data to retrieve about the stands from the database, and in what format to return it (for example, in a table). 

-------------------

In the code, I created:

1. @extend_schema and @action decorators for the endpoint method
2. endpoint method stand_treatment_results, which (a) recovers & validates the stand id, (b) recovers the treatment plan id, (c) feeds both to the services.py function to get treatment stand results, and (d) serializes and returns the result to the user.
3. the shell of the get_treatment_result_for_stand function in impacts/services.py

Foxbyte0101 commented:
@george-silva @roquebetioljr Made progress on the endpoint, namely making a custom serializer class in impacts/serializers.py, and building out the retrieval logic for the table in impacts/services.py. Also changed the code in impacts/views.py to work with these other parts of the impacts app. I was able to get black formatter to work which is good (passed the linter test).

Next step is to make sure all parts of the endpoint are referencing correct pieces, and then testing the endpoint with unit tests.

roquebetioljr commented on src/planscape/impacts/services.py:
```diff
@@ -528,3 +528,58 @@ def calculate_project_area_deltas(
+
+def get_treatment_results_table_data(treatment_plan, stand_id):
+    """
+    Retrieves treatment results for RATE_OF_SPREAD and FLAME_LENGTH for the given stand and treatment plan.
+    Return a list of dictionaries: [{"year": ..., "rate_of_spread": ..., "flame_length": ...}, ...]
+    """
+    # Filter results for the given stand and plan, for the two variables we need
+    queryset = TreatmentResult.objects.filter(
+        treatment_plan=treatment_plan,
+        stand_id=stand_id,
+        variable__in=[ImpactVariable.RATE_OF_SPREAD, ImpactVariable.FLAME_LENGTH],
+    )
```
You can let the front-end decide what needs to be filtered.

Also, you can simplify it by returning one object per result. Ex: `[{"year": 2024, "variable": "FL", "value": "High"},...]`
This way, you can execute a query instead of iterate over the result and reducing the number of queries.
```suggestion
    Retrieves treatment results for RATE_OF_SPREAD and FLAME_LENGTH for the given stand and treatment plan.
    Return a list of dictionaries: [{"year": ..., "rate_of_spread": ..., "flame_length": ...}, ...]
    """
    filters = {}
    if variables:
        filters = {"variable__in": variables}
    # Filter results for the given stand and plan, for the two variables we need
    queryset = TreatmentResult.objects.filter(
        treatment_plan=treatment_plan,
        stand_id=stand_id,
        **filters
    )
    result = queryset.values("year", "variable", "value") 
```

roquebetioljr replied:
Also, another way, and probably better, is to return the filtered queryset and build the return with a serializer in the viewset.

Foxbyte0101 commented:
@george-silva --changed the services.py code to include all variables in TreatmentResult model, and to only query the database once, creating a dictionary of the result, and then a table.

Foxbyte0101 commented:
@george-silva, added if-then logic for flame length and rate of spread relative values, as well as other changes to services.py. General flow is views.py for endpoint creation, serializer.py for validation, and services.py for data retrieval and table creation.

Foxbyte0101 commented:
@george-silva, removed most comments, removed if-not code in views.py, and refactored services.py code, including using non-abbreviated variable names & values. Should be easier to read now!

pflopez commented on src/planscape/impacts/services.py:
```diff
@@ -528,3 +533,96 @@ def calculate_project_area_deltas(
+    These cut-off boundaries match BehavePlus6 spreadsheet on GD.
+    """
+    if ros_value < 3.0:
+        return "VERY_LOW"
```
yeah good call, given that we are returning the human readable values here, might as well be the values we'll display directly

Foxbyte0101 commented:
@george-silva @roquebetioljr, added baseline and changed the string to human readable text. Also added the progress on the unit tests for the stand info endpoint's views.py, serializers.py, and services.py. I added the tests in impacts/tests/test_services.py & test_views.py. The tests pass, but making sure they're the right tests/any edge cases need to be added.

Foxbyte0101 commented:
Cleared up conflicts with main, and fixed unit test for test_views.py. Now all four unit tests for my views.py code pass. All the imports are organized using isort, & black formatter was used throughout.
